components/script/dom/documentorshadowroot.rs

```diff
@@ -0,0 +1,272 @@
+        let x = *x as f32;
+        let y = *y as f32;
+        let point = &Point2D::new(x, y);
+        let viewport = self.window.window_size().initial_viewport;
+
+        if !has_browsing_context {
+            return vec![];
+        }
+
+        // Step 2
+        if x < 0.0 || y < 0.0 || x > viewport.width || y > viewport.height {
+            return vec![];
+        }
+
+        let js_runtime = unsafe { JS_GetRuntime(self.window.get_cx()) };
+
+        // Step 1 and Step 3
+        let nodes = self.nodes_from_point(point, NodesFromPointQueryType::All);
+        let mut elements: Vec<DomRoot<Element>> = nodes
+            .iter()
+            .flat_map(|&untrusted_node_address| {
+                let node = unsafe {
+                    node::from_untrusted_node_address(js_runtime, untrusted_node_address)
+                };
+                DomRoot::downcast::<Element>(node)
+            })
+            .collect();
+
+        // Step 4
+        if let Some(root_element) = document_element {
+            if elements.last() != Some(&root_element) {
+                elements.push(root_element);
+            }
+        }
+
+        // Step 5
+        elements
+    }
+
+    // https://html.spec.whatwg.org/multipage/#dom-document-activeelement
+    pub fn get_active_element(
+        &self,
+        focused_element: Option<DomRoot<Element>>,
+        body: Option<DomRoot<HTMLElement>>,
+        document_element: Option<DomRoot<Element>>,
+    ) -> Option<DomRoot<Element>> {
+        // TODO: Step 2.
+
+        match focused_element {
+            Some(element) => Some(element), // Step 3. and 4.
+            None => match body {
+                // Step 5.
+                Some(body) => Some(DomRoot::upcast(body)),
+                None => document_element,
+            },
+        }
+    }
+
+    /// Remove a stylesheet owned by `owner` from the list of document sheets.
+    #[allow(unrooted_must_root)] // Owner needs to be rooted already necessarily.
+    pub fn remove_stylesheet(
+        &self,
+        owner: &Element,
+        s: &Arc<Stylesheet>,
+        stylesheets: &mut StylesheetSet<StyleSheetInDocument>,
+    ) {
+        self.window
+            .layout_chan()
+            .send(Msg::RemoveStylesheet(s.clone()))
+            .unwrap();
+
+        let guard = s.shared_lock.read();
+
+        // FIXME(emilio): Would be nice to remove the clone, etc.
+        stylesheets.remove_stylesheet(
+            None,
+            StyleSheetInDocument {
+                sheet: s.clone(),
+                owner: Dom::from_ref(owner),
+            },
+            &guard,
+        );
+    }
+
+    /// Add a stylesheet owned by `owner` to the list of document sheets, in the
+    /// correct tree position.
+    #[allow(unrooted_must_root)] // Owner needs to be rooted already necessarily.
+    pub fn add_stylesheet(
+        &self,
+        owner: &Element,
+        stylesheets: &mut StylesheetSet<StyleSheetInDocument>,
```

Comment:
It's a bit unfortunate to use virtual dispatch here...

Also this doesn't seem to use `self` at all.